File: Leetcode/Array/152.py
class Solution:
    def maxProduct(self, nums: List[int]) -> int:
        max_product, min_product, answer = nums[0], nums[0], nums[0]
        for i in range(1, len(nums)):
            x = max(nums[i], max_product * nums[i], min_product * nums[i])
            y = min(nums[i], max_product * nums[i], min_product * nums[i])
            max_product, min_product = x, y
            answer = max(max_product, answer)

        return answer

File: Leetcode/Array/152.js
/**
 * @param {number[]} nums
 * @return {number}
 */
var maxProduct = function (nums) {
  if (nums.length === 1) return nums[0];

  let maxProd = nums[0];
  let minProd = nums[0];
  let answer = nums[0];

  for (let i = 1; i < nums.length; i++) {
    let tempMax = Math.max(maxProd * nums[i], minProd * nums[i], nums[i]);
    let tempMin = Math.min(maxProd * nums[i], minProd * nums[i], nums[i]);
    maxProd = tempMax;
    minProd = tempMin;
    answer = Math.max(maxProd, answer);
  }

  return answer;
};
